src/main/scala/sparkz/core/network/peer/PeerBucketStorage.scala

```diff
@@ -0,0 +1,196 @@
+package sparkz.core.network.peer
+
+import com.google.common.primitives.{Bytes, Ints}
+import scorex.crypto.hash.Blake2b256
+import sparkz.core.network.peer.BucketManager.PeerBucketValue
+import sparkz.core.network.peer.PeerBucketStorage.{BucketConfig, BucketHashContent, NOT_FOUND_PEER_INDEXES}
+import sparkz.core.utils.TimeProvider.Time
+import sparkz.core.utils.{NetworkAddressWrapper, TimeProvider}
+
+import java.net.InetSocketAddress
+import java.nio.ByteBuffer
+import scala.collection.mutable
+import scala.util.{Success, Try}
+
+abstract class PeerBucketStorage[T <: BucketHashContent](
+                                  private val bucketConfig: BucketConfig,
+                                  private val nKey: Int,
+                                  private val timeProvider: TimeProvider,
+                                  private val isNewPeer: Boolean
+                                ) {
+  private val (buckets, positionsInBucket, bucketSubgroup) = BucketConfig.unapply(bucketConfig).getOrElse(throw new IllegalArgumentException())
+  private val table: mutable.Map[(Int, Int), (PeerBucketValue, Time)] = mutable.Map.empty[(Int, Int), (PeerBucketValue, Time)]
+  private val reverseTable: mutable.Map[InetSocketAddress, (Int, Int)] = mutable.Map.empty[InetSocketAddress, (Int, Int)]
+
+  protected[peer] def getBucket(hashContent: T): Int = {
+    val hash1Content = Bytes.concat(Ints.toByteArray(nKey), hashContent.getHashAsBytes)
+    val h: Array[Byte] = Blake2b256.hash(hash1Content)
+    val hInt = ByteBuffer.wrap(h).getInt
+
+    val hash2Content = Bytes.concat(Ints.toByteArray(nKey), getAddressGroup(hashContent.getAddress), Ints.toByteArray(hInt % bucketSubgroup))
+    val hash2: Array[Byte] = Blake2b256.hash(hash2Content)
+
+    ByteBuffer.wrap(hash2).getInt % buckets
+  }
+
+  private def getAddressGroup(peerAddress: InetSocketAddress): Array[Byte] = {
+    val na = new NetworkAddressWrapper(peerAddress)
+    var nClass = NetworkAddressWrapper.NET_IPV6
+    var nStartByte = 0
+    var nBits = 16
```

Comment:
the value of nBits is only used to count while loop cycles and effectively has 3 values:
- 16 - default - 1 cycle
- 0 - 0 cycles
- 32\36 - 2 cycles

I struggle to understand what this code does, can you please add a comment for this method or rename some vars to be more descriptive?

Reply:
Sure, I can add some comments and change the variables' names to make it clearer. This method should create a byte array for every address' subgroup based on different protocols